### DISCUSSION TITLE
ValueError on nlp.set_kb() with SpaCy v3

### DISCUSSION THREAD

**Discussion by bmiret-teklia** [explosion/spaCy#7853] [Help: Coding & Implementations] [answered]
Hi,

I tried to follow the tutorial on Entity Linker training but I'm stuck at the training part with the following error when calling `set_kb()`: 

```
ValueError: [E885] entity_linker.set_kb received an invalid 'kb_loader' argument: expected a callable function, but got: <class 'spacy.kb.KnowledgeBase'>

```
The same code runs fine on spaCy 2.2.4 (with the "v2 version" lines). I tried to look at what changed in spaCy v3 but didn't see anything that could help (but real beginner here). Also, EntityLinker.set_kb is tagged as a new v3 feature in the documentation...?
Did I miss something?

## How to reproduce the behaviour

```
import spacy
from spacy.kb import KnowledgeBase

nlp = spacy.load("en_core_web_sm")

name_dict ={
    'Q312545': 'Roy Stanley Emerson',
    'Q48226': 'Ralph Waldo Emerson',
    'Q215952': 'Emerson Ferreira da Rosa'
}

desc_dict = {
    'Q312545': 'Australian tennis player',
    'Q48226': 'American philosopher, essayist, and poet',
    'Q215952': 'Brazilian footballer'
}

kb = KnowledgeBase(vocab=nlp.vocab, entity_vector_length=96)

for qid, desc in desc_dict.items():
    desc_doc = nlp(desc)
    desc_enc = desc_doc.vector
    kb.add_entity(entity=qid, entity_vector=desc_enc, freq=324)

for qid, name in name_dict.items():
    kb.add_alias(alias=name, entities=[qid], probabilities=[1])

qids = name_dict.keys()
probs=[0.3 for qid in qids]
kb.add_alias(alias="Emerson", entities=qids, probabilities=probs)

# v3 version
entity_linker = nlp.add_pipe("entity_linker", config={"incl_prior":False}, last=True)
entity_linker.set_kb(kb)

# # v2 version
# entity_linker = nlp.create_pipe("entity_linker", config={"incl_prior":False})
# entity_linker.set_kb(kb)
#  nlp.add_pipe(entity_linker, last=True)
```
## Info about spaCy

- **spaCy version:** 3.0.5
- **Platform:** Linux-4.15.0-136-generic-x86_64-with-Ubuntu-18.04-bionic
- **Python version:** 3.6.9
- **Pipelines:** en_core_web_sm (3.0.0)

Thanks!

**Comment by svlandeg** [comment]
Hi! As a maintenance note: this is a good question for the discussions forum, so I'll move it there. The original issue will be closed and will link through to the open discussion thread.

**Comment by svlandeg** [comment, marked as answer]
The `entity_linker` functionality was considered experimental in v2, which is why it's marked as "new" in the docs. Some of the API details have indeed stabilized since then.

In a nutshell, in spaCy v3 we've introduced [training from config file](https://spacy.io/usage/training#config), which means that the config file is a single source of truth and contains all your (hyper)parameters to train a pipeline. It's very powerful and extensible, and works with a [registry](https://spacy.io/api/top-level#registry) where you can find predefined functions to use in your config, or create and register your own.

This is also the reason why [`entity_linker.set_kb`](https://spacy.io/api/entitylinker) now takes a function instead of the KB object itself. For instance, something like this:
```
@registry.misc("my_custom_kb_loader")
def create_kb(vocab):
    kb = KnowledgeBase(vocab, entity_vector_length=128)
    kb.add_entity(...)
    kb.add_alias(...)
    return kb
entity_linker = nlp.add_pipe("entity_linker")
entity_linker.set_kb(create_kb)
```
In the config file, you would specify the `kb_loader` in the initialization block:
```
[initialize.components.entity_linker.kb_loader]
@misc = "my_custom_kb_loader"
```
You can find the built-in architectures here: https://spacy.io/api/architectures#entitylinker and it's probably useful to check out this example project as well: https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson.

Hope that helps!

**Reply by bmiret-teklia** [reply]
The example project is a great help indeed and I'm now starting to get my head around the config file and registry system. Thanks a lot for the quick and patient reply!
